File: pa3/Dictionary.java
//----------------------------------------------------------------------
// Dictionary.java
//----------------------------------------------------------------------

public class Dictionary implements DictionaryInterface {
	
    //private inner Node class
    private class Node {
    	String key;
        String value;
        Node next;
        
        Node(String key, String value){
        	this.key = key; 	//set the new key as the class's key
        	this.value = value; //set the new value as the class's value
        	next = null;
        }
    }

    //fields for the Dictionary class
    private Node head;		//reference to first Node in List
    private int numItems;   //number of items in the Dictionary

    //constructor for the Dictionary class
    public Dictionary(){
    	head = null;
        numItems = 0;
    }


    //private helper function -----------------------------------------

    //findKey()
    //return a reference to Node containing its argument key,
    //or return null if no such Node exit.
    private Node findKey(String key){
    	Node n = head;
    	
        while(n != null){
        	if(n.key != key){
        		n = n.next;
    	    } else{
    	    	//System.out.println("in findKey");
    	    	return n;
    	    }
        }
        return null; //no such Node exist
    }

    
    //ADT operations --------------------------------------------------
    
    //isEmpty()
    //returns true if the Dictionary contains no pairs, false otherwise
    public boolean isEmpty(){
    	return(numItems == 0);
    }

    //size()
    //returns the number of (key, value) pairs in the Dictionary
    public int size() {
    	return numItems;
    }
    
    //lookup()
    //returns the value field when the Dictionary contains a matching pair of key and argument key
    public String lookup(String key){
    	Node n = findKey(key);
	    while(n != null){
	    	if(n.key == key){
	    		//System.out.println("in lookup");
	    		return n.value;
	    	}

		    n = n.next;
	    }
	    //System.out.println("out of lookup if loop, returning null");
	    return null; //return null if can't find it
    }
    
    //insert()
    //precondition: Dictionary does NOT currently contain the argument key
    public void insert(String key, String value) throws DuplicateKeyException{
    	if(lookup(key) == null){ //new unused key
    		if(head == null){ //adding the new (key, value) as the head when the Node is empty
    			head = new Node(key, value);
    			numItems++;
    			
    		}else{ //(head != null)
    			Node n = head;//start at head
    			
    			while(n.next != null){//while loop until n.next is going to be null
    				n = n.next; 
    			}
    			n.next = new Node(key, value);//inserting the (key, value) at the end
    			numItems++;
    		}
    	} else{ //throw exception when a duplicate key is being used
    		throw new DuplicateKeyException("cannot insert duplicate keys");
    	}
    }
    
    //delete()
    //precondition: Dictionary currently contains the argument key
    public void delete(String key) throws KeyNotFoundException{
    	if(lookup(key) == null){ //throw exception when trying to delete a non-existing key
    		throw new KeyNotFoundException("cannot delete non-existent key");
    	} else{ //found the key that is going to be deleted
    		Node n = head;
    		boolean keyFound = false; //exit the loop when key is found
    		while(n != null && !keyFound){ //while the key is NOT found
    			if(numItems == 1 && head.key==key){//when there's only 1 item
    				head = null;
    				numItems--; //decrease numItems to 0
    				keyFound = true;//exit
    			} else if(n.next.key == key){
    				n.next = n.next.next;
    				numItems--;
    				keyFound = true; //exit
    			} else if(n.key == key){
    				head = n.next;
    				keyFound = true;
    				numItems--;
    			} else n = n.next;
    		}
    	}
    }

    //makeEmpty()
    //Resets the Dictionary to the empty state 
    public void makeEmpty(){
    	head = null;
    	numItems = 0;
    }

    //toString()
    public String toString(){
    	Node n = head; //set n as the head of the Node
    	String s = ""; //initialize string s
    	
    	while(n != null){
    		s += n.key + " " + n.value + "\n"; //adding key and value to the string
    		n = n.next;
    	}
    	return s; //returning the current state of the Dictionary
    }
}

File: 5 - C Dictionary ADT/Dictionary.c
/*
 * Annie Shen (ashen7@ucsc.edu)
 * Dictionary.c
 * Implementation file for Dictionary ADT
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"

//private types -------------------------------------------------------

//NodeObj
typedef struct NodeObj{
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

//Node
typedef NodeObj* Node;

//newNode()
//constructor for Node type
Node newNode(char* k, char* v) {
    Node N = malloc(sizeof(NodeObj));
    assert(N!=NULL); //die if N equals to NULL
    N->key = k;
    N->value = v;
    N->next = NULL;
    return(N);
}

//freeNode()
//destructor for the Node type
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
        free(*pN); //prevent memory leak
        *pN = NULL;
    }
}

//DictionaryObj
typedef struct DictionaryObj{
   Node top;
   int numItems;
} DictionaryObj;


//public functions -----------------------------------------------

//newDictionary()
//constructor for the Dictionary type
Dictionary newDictionary(void){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->top = NULL;
   D->numItems = 0;
   return D;
}

//freeDictionary()
//destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
       if( !isEmpty(*pD) ){
           makeEmpty(*pD);
       }
       free(*pD);
       *pD = NULL;
   }
}

//isEmpty()
//returns 1 (true) if D is empty, 0 (false) otherwise
int isEmpty(Dictionary D) {
    if (D == NULL) { //when Dictionary is NULL
        fprintf(stderr, "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
    } else if(D->numItems == 0) {
        return 1; //return true because it is empty
    } else{
        return 0; //return false saying it is NOT empty
    }
}

//size()
//returns the size of (key,value) pair in D
int size(Dictionary D){
    return D->numItems;
}

//lookup()
//return the value that is in D's (key, value), else return NULL if there's no such key
char* lookup(Dictionary D, char* k){
    Node N = D->top;

    while (N != NULL){
        if(strcmp(N->key, k) == 0){ //(=0) means true, (=1) equals false
            //printf("lookup: %s\n",N->value);
            return N->value;
        }
        N = N->next;
    }
    //printf("null\n");
    return NULL; //return NULL if there's no such pair
}

//insert()
//insert new (key, value) into D
void insert(Dictionary D, char* k, char* v) {
    //considering if D is empty or already had contents
    Node N = newNode(k, v); //new Node

    if (lookup(D, k) == NULL) { //the pair doesn't exist in D yet
            N->next = D->top;
            D->top = N;
            D->numItems++;
        }
     else{ //the key is duplicate!
        fprintf(stderr, "Dictionary error: cannot insert duplicate keys\n");
        exit(EXIT_FAILURE);
    }
    //Forgot this is Stack, not Queue!
//    Node top = D->top;
//    Node new = newNode(k, v);
//
//    if(lookup(D, k) == NULL) { //if the pair doesn't exist
//        if(top == NULL) { //if D is empty
//            top = new;
//        } else { //if D is NOT empty
//            if(top->next != NULL) {
//                top = top->next;
//            }
//            top->next = new;
//        }
//        D->numItems++; //now one more item
//    } else{ //as in if the key existed
//        fprintf(stdout, "Dictionary error: cannot insert duplicate keys");
//    }
}

//delete()
//delete k key from D's (key, value)
void delete(Dictionary D, char* k){
    Node N = D->top;

    if(lookup(D, k) == NULL){
        fprintf(stderr, "Dictionary error: cannot delete non-existing keys\n");
        exit(EXIT_FAILURE);
    }

    //If the key is found! then:
    if(strcmp(N->key, k) == 0){ //if the first node matches the key
        D->top = N->next;
        freeNode(&N); //prevent memory leak
        N = NULL; //initialize
        D->numItems--;

    } else{ //if it's after the first node
        //int found = 0;
        while(N != NULL && N->next != NULL){
            if(strcmp(N->next->key, k) == 0){//(=0) means true, (=1) equals false
                Node G = N;
                Node C = N->next;
                G->next = C->next;
                N = G;
                freeNode(&C);
                //Node temp = N->next; //set M as N
                //N->next = N->next->next; //set temp as N's next key
                //freeNode(&temp); //discard temp
                //temp = NULL;
                D->numItems--;
                //found =1;
            }
            N = N->next; //loop until key is found
        }

    }
}

void empty(Node N){
    if(N != NULL){
        empty(N->next);
        freeNode(&N);
    }
}

//makeEmpty()
//resets-empty Dictionary D
void makeEmpty(Dictionary D){
    empty(D->top);
    D->top = NULL;
    D->numItems = 0;
}

//printRec()
//additional help for printDictionary() to print recursively
void printRec(FILE* out, Node N){
    //printf("in printRec, outside of while loop");
    if(N != NULL) {
        printRec(out, N->next);
        //printf("in printRec");
        fprintf(out, "%s %s\n", N->key, N->value);
    }
}

//printDictionary()
//prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
    //Node N;

    if(D == NULL){
        fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    //printf("in printDictionary, in loop before calling printRec\n");
    printRec(out, D->top);

//    for(N=D->top; N!=NULL; N=N->next){
//        fprintf(out, "%s %s\n", N->key, N->value);
//    }
}
